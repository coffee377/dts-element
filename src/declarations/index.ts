export * from './class-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './interface-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
