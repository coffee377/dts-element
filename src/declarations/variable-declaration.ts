import * as ts from 'typescript';
import {IType} from '../collections';
import {any_type, ElementKind} from '../constants';
import {create_element, IElement} from '../element';
import {transform} from '../transform';

export interface IVariableDeclarationOptions {
  name: string;
  type?: IType;
  export?: boolean;
  const?: boolean;
  let?: boolean;
}

export interface IVariableDeclaration
  extends IElement<ElementKind.VariableDeclaration>, IVariableDeclarationOptions {}

export const create_variable_declaration = (options: IVariableDeclarationOptions): IVariableDeclaration => ({
  ...create_element(ElementKind.VariableDeclaration),
  ...options,
});

// tslint:disable:ter-indent

export const transform_variable_declaration = (element: IVariableDeclaration, path: IElement<any>[]) => {
  const type = transform(element.type || any_type, path) as ts.TypeNode;
  return ts.createVariableStatement(
    /* modifiers       */ (element.export === true)
                            ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
                            : undefined,
    /* declarationList */ ts.createVariableDeclarationList(
                            /* declarations  */ [ts.createVariableDeclaration(element.name, type)],
                            /* flags         */ (element.const === true)
                                                  ? ts.NodeFlags.Const
                                                  : (element.let === true)
                                                    ? ts.NodeFlags.Let
                                                    : undefined,
                          ),
  );
};
