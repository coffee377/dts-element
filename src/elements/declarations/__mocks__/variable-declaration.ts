import {Declaration} from '../../declaration';

export class VariableDeclaration extends Declaration<any, any> {

  public get default_parameters(): any {
    return super.default_declaration_parameters;
  }

  public _emit_raw(): string {
    return `[VariableDeclaration ${this.parameters.name}]`;
  }

}
