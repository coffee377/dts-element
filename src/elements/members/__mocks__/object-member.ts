import {Member} from '../../member';

export class ObjectMember extends Member<any, any> {

  public get default_parameters(): any {
    return {};
  }

  public _emit(): string {
    return `[ObjectMember ${this.parameters.owned.emit()}]`;
  }

}
