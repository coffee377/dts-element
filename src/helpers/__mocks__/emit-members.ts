export const emit_members = (): string => '[emit-members]';
