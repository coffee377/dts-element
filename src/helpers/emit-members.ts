import {Container} from '../collections';
import {Member} from '../elements/member';
import {emit_elements} from './emit-elements';
import {indent_every_line} from './indent-every-line';

export const emit_members = (children: Member[], container: Container): string =>
  (children.length === 0)
    ? '{}'
    : `{\n${indent_every_line(emit_elements(children, container))}\n}`;
