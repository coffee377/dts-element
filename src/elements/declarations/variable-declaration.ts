import {Container} from '../../collections';
import {any_type} from '../../constants';
import {emit_declaration_name} from '../../helpers/emit-declaration-name';
import {emit_declare} from '../../helpers/emit-declare';
import {emit_optional} from '../../helpers/emit-optional';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {Type} from '../type';

export type VariableKind = 'var' | 'let' | 'const';

// tslint:disable-next-line no-empty-interface
export interface IVariableDeclarationRequiredParameters {}

export interface IVariableDeclarationOptionalParameters {
  kind: VariableKind;
  type: Type;
  optional: boolean;
}

export class VariableDeclaration
    extends Declaration<IVariableDeclarationRequiredParameters, IVariableDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IVariableDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      kind: 'var' as VariableKind,
      type: any_type,
      optional: false,
    });
  }

  public _emit_raw(container: Container): string {
    const {kind, type} = this.parameters;
    const name = emit_declaration_name(this.parameters.name, container);
    const optional = emit_optional(this.parameters.optional);
    const declare = emit_declare(container);
    return `${declare}${kind} ${name}${optional}: ${type._emit(this)};`;
  }

}
