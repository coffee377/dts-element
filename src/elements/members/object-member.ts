import {emit_readonly} from '../../helpers/emit-readonly';
import {Stack} from '../../stack';
import {Member} from '../member';

// tslint:disable-next-line no-empty-interface
export interface IObjectMemberRequiredParameters {}

export interface IObjectMemberOptionalParameters {
  readonly: boolean;
}

export class ObjectMember extends Member<IObjectMemberRequiredParameters, IObjectMemberOptionalParameters> {

  public get default_parameters(): IObjectMemberOptionalParameters {
    return {
      readonly: false,
    };
  }

  public _emit(stack: Stack): string {
    const {owned} = this.parameters;
    const readonly = emit_readonly(this.parameters.readonly, owned);
    return `${owned._emit_jsdoc()}${readonly}${owned._emit_raw(stack.push(owned))}`;
  }

}
