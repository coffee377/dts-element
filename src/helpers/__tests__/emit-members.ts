jest.unmock('../emit-members');

import {VariableDeclaration} from '../../elements/declarations/variable-declaration';
import {ObjectMember} from '../../elements/members/object-member';
import {emit_members} from '../emit-members';

it('should return empty object while there are no children', () => {
  expect(emit_members([], null)).toBe('{}');
});

it('should return indented members while there are children', () => {
  expect(emit_members([
    new ObjectMember({owned: new VariableDeclaration({name: 'x'})}),
    new ObjectMember({owned: new VariableDeclaration({name: 'y'})}),
  ], null)).toMatchSnapshot();
});
