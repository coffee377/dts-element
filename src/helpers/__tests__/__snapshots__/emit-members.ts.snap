// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return indented members while there are children 1`] = `
"{
[indent-every-line
  [emit-elements
    [ObjectMember [emit-jsdoc][VariableDeclaration x]]
    [ObjectMember [emit-jsdoc][VariableDeclaration y]]
  ]
]
}"
`;
